src/generic/GenericTypes.jl

```diff
@@ -319,19 +319,21 @@ mutable struct PolyRing{T <: RingElement} <: AbstractAlgebra.PolyRing{T}
    S::Symbol
 
    function PolyRing{T}(R::Ring, s::Symbol, cached::Bool = true) where T <: RingElement
-      if cached && haskey(PolyID, (R, s))
-         return PolyID[R, s]::PolyRing{T}
-      else
-         z = new{T}(R, s)
-         if cached
-           PolyID[R, s] = z
+      if cached
+         S = get(PolyID, (R, s), nothing)
+         if S !== nothing
+            return S::PolyRing{T}
```

Comment:
Just some thoughts, feel free to ignore:

So what you call here `S` is the same as what the old code / rest of the function call `z`, right?

Since this pattern repeats a lot, we could put it into a separate helper, simplified by using the `get!` method:
```julia
function get_cached!(default::Base.Callable, dict::CacheDictType, key, use_cache::Bool)
   return use_cache ? get!(default, dict, key) : default()
end
```
Then this constructor could become something like this:
```julia
function PolyRing{T}(R::Ring, s::Symbol, cached::Bool = true) where T <: RingElement
   return get_cached!(PolyID, (R, s), cached) do
      new{T}(R, s)
   end
end
```